README
NAME
    Coro - the only real threads in perl

SYNOPSIS
      use Coro;
  
      async {
         # some asynchronous thread of execution
         print "2\n";
         cede; # yield back to main
         print "4\n";
      };
      print "1\n";
      cede; # yield to coro
      print "3\n";
      cede; # and again
  
      # use locking
      use Coro::Semaphore;
      my $lock = new Coro::Semaphore;
      my $locked;
  
      $lock->down;
      $locked = 1;
      $lock->up;

DESCRIPTION
    For a tutorial-style introduction, please read the Coro::Intro manpage.
    This manpage mainly contains reference information.

    This module collection manages continuations in general, most often in
    the form of cooperative threads (also called coros, or simply "coro" in
    the documentation). They are similar to kernel threads but don't (in
    general) run in parallel at the same time even on SMP machines. The
    specific flavor of thread offered by this module also guarantees you
    that it will not switch between threads unless necessary, at
    easily-identified points in your program, so locking and parallel access
    are rarely an issue, making thread programming much safer and easier
    than using other thread models.

    Unlike the so-called "Perl threads" (which are not actually real threads
    but only the windows process emulation ported to unix, and as such act
    as processes), Coro provides a full shared address space, which makes
    communication between threads very easy. And Coro's threads are fast,
    too: disabling the Windows process emulation code in your perl and using
    Coro can easily result in a two to four times speed increase for your
    programs. A parallel matrix multiplication benchmark runs over 300 times
    faster on a single core than perl's pseudo-threads on a quad core using
    all four cores.

    Coro achieves that by supporting multiple running interpreters that
    share data, which is especially useful to code pseudo-parallel processes
    and for event-based programming, such as multiple HTTP-GET requests
    running concurrently. See Coro::AnyEvent to learn more on how to
    integrate Coro into an event-based environment.

    In this module, a thread is defined as "callchain + lexical variables +
    some package variables + C stack), that is, a thread has its own
    callchain, its own set of lexicals and its own set of perls most
    important global variables (see Coro::State for more configuration and
    background info).

    See also the "SEE ALSO" section at the end of this document - the Coro
    module family is quite large.

GLOBAL VARIABLES
    $Coro::main
        This variable stores the Coro object that represents the main
        program. While you cna "ready" it and do most other things you can
        do to coro, it is mainly useful to compare again $Coro::current, to
        see whether you are running in the main program or not.

    $Coro::current
        The Coro object representing the current coro (the last coro that
        the Coro scheduler switched to). The initial value is $Coro::main
        (of course).

        This variable is strictly *read-only*. You can take copies of the
        value stored in it and use it as any other Coro object, but you must
        not otherwise modify the variable itself.

    $Coro::idle
        This variable is mainly useful to integrate Coro into event loops.
        It is usually better to rely on Coro::AnyEvent or Coro::EV, as this
        is pretty low-level functionality.

        This variable stores either a Coro object or a callback.

        If it is a callback, the it is called whenever the scheduler finds
        no ready coros to run. The default implementation prints "FATAL:
        deadlock detected" and exits, because the program has no other way
        to continue.

        If it is a coro object, then this object will be readied (without
        invoking any ready hooks, however) when the scheduler finds no other
        ready coros to run.

        This hook is overwritten by modules such as "Coro::EV" and
        "Coro::AnyEvent" to wait on an external event that hopefully wake up
        a coro so the scheduler can run it.

        Note that the callback *must not*, under any circumstances, block
        the current coro. Normally, this is achieved by having an "idle
        coro" that calls the event loop and then blocks again, and then
        readying that coro in the idle handler, or by simply placing the
        idle coro in this variable.

        See Coro::Event or Coro::AnyEvent for examples of using this
        technique.

        Please note that if your callback recursively invokes perl (e.g. for
        event handlers), then it must be prepared to be called recursively
        itself.

SIMPLE CORO CREATION
    async { ... } [@args...]
        Create a new coro and return its Coro object (usually unused). The
        coro will be put into the ready queue, so it will start running
        automatically on the next scheduler run.

        The first argument is a codeblock/closure that should be executed in
        the coro. When it returns argument returns the coro is automatically
        terminated.

        The remaining arguments are passed as arguments to the closure.

        See the "Coro::State::new" constructor for info about the coro
        environment in which coro are executed.

        Calling "exit" in a coro will do the same as calling exit outside
        the coro. Likewise, when the coro dies, the program will exit, just
        as it would in the main program.

        If you do not want that, you can provide a default "die" handler, or
        simply avoid dieing (by use of "eval").

        Example: Create a new coro that just prints its arguments.

           async {
              print "@_\n";
           } 1,2,3,4;

    async_pool { ... } [@args...]
        Similar to "async", but uses a coro pool, so you should not call
        terminate or join on it (although you are allowed to), and you get a
        coro that might have executed other code already (which can be good
        or bad :).

        On the plus side, this function is about twice as fast as creating
        (and destroying) a completely new coro, so if you need a lot of
        generic coros in quick successsion, use "async_pool", not "async".

        The code block is executed in an "eval" context and a warning will
        be issued in case of an exception instead of terminating the
        program, as "async" does. As the coro is being reused, stuff like
        "on_destroy" will not work in the expected way, unless you call
        terminate or cancel, which somehow defeats the purpose of pooling
        (but is fine in the exceptional case).

        The priority will be reset to 0 after each run, tracing will be
        disabled, the description will be reset and the default output
        filehandle gets restored, so you can change all these. Otherwise the
        coro will be re-used "as-is": most notably if you change other
        per-coro global stuff such as $/ you *must needs* revert that
        change, which is most simply done by using local as in: "local $/".

        The idle pool size is limited to 8 idle coros (this can be adjusted
        by changing $Coro::POOL_SIZE), but there can be as many non-idle
        coros as required.

        If you are concerned about pooled coros growing a lot because a
        single "async_pool" used a lot of stackspace you can e.g.
        "async_pool { terminate }" once per second or so to slowly replenish
        the pool. In addition to that, when the stacks used by a handler
        grows larger than 32kb (adjustable via $Coro::POOL_RSS) it will also
        be destroyed.

STATIC METHODS
    Static methods are actually functions that implicitly operate on the
    current coro.

    schedule
        Calls the scheduler. The scheduler will find the next coro that is
        to be run from the ready queue and switches to it. The next coro to
        be run is simply the one with the highest priority that is longest
        in its ready queue. If there is no coro ready, it will clal the
        $Coro::idle hook.

        Please note that the current coro will *not* be put into the ready
        queue, so calling this function usually means you will never be
        called again unless something else (e.g. an event handler) calls
        "->ready", thus waking you up.

        This makes "schedule" *the* generic method to use to block the
        current coro and wait for events: first you remember the current
        coro in a variable, then arrange for some callback of yours to call
        "->ready" on that once some event happens, and last you call
        "schedule" to put yourself to sleep. Note that a lot of things can
        wake your coro up, so you need to check whether the event indeed
        happened, e.g. by storing the status in a variable.

        See HOW TO WAIT FOR A CALLBACK, below, for some ways to wait for
        callbacks.

    cede
        "Cede" to other coros. This function puts the current coro into the
        ready queue and calls "schedule", which has the effect of giving up
        the current "timeslice" to other coros of the same or higher
        priority. Once your coro gets its turn again it will automatically
        be resumed.

        This function is often called "yield" in other languages.

    Coro::cede_notself
        Works like cede, but is not exported by default and will cede to
        *any* coro, regardless of priority. This is useful sometimes to
        ensure progress is made.

    terminate [arg...]
        Terminates the current coro with the given status values (see
        cancel).

    Coro::on_enter BLOCK, Coro::on_leave BLOCK
        These function install enter and leave winders in the current scope.
        The enter block will be executed when on_enter is called and
        whenever the current coro is re-entered by the scheduler, while the
        leave block is executed whenever the current coro is blocked by the
        scheduler, and also when the containing scope is exited (by whatever
        means, be it exit, die, last etc.).

        *Neither invoking the scheduler, nor exceptions, are allowed within
        those BLOCKs*. That means: do not even think about calling "die"
        without an eval, and do not even think of entering the scheduler in
        any way.

        Since both BLOCKs are tied to the current scope, they will
        automatically be removed when the current scope exits.

        These functions implement the same concept as "dynamic-wind" in
        scheme does, and are useful when you want to localise some resource
        to a specific coro.

        They slow down thread switching considerably for coros that use them
        (about 40% for a BLOCK with a single assignment, so thread switching
        is still reasonably fast if the handlers are fast).

        These functions are best understood by an example: The following
        function will change the current timezone to
        "Antarctica/South_Pole", which requires a call to "tzset", but by
        using "on_enter" and "on_leave", which remember/change the current
        timezone and restore the previous value, respectively, the timezone
        is only changed for the coro that installed those handlers.

           use POSIX qw(tzset);

           async {
              my $old_tz; # store outside TZ value here

              Coro::on_enter {
                 $old_tz = $ENV{TZ}; # remember the old value

                 $ENV{TZ} = "Antarctica/South_Pole";
                 tzset; # enable new value
              };

              Coro::on_leave {
                 $ENV{TZ} = $old_tz;
                 tzset; # restore old value
              };

              # at this place, the timezone is Antarctica/South_Pole,
              # without disturbing the TZ of any other coro.
           };

        This can be used to localise about any resource (locale, uid,
        current working directory etc.) to a block, despite the existance of
        other coros.

        Another interesting example implements time-sliced multitasking
        using interval timers (this could obviously be optimised, but does
        the job):

           # "timeslice" the given block
           sub timeslice(&) {
              use Time::HiRes ();

              Coro::on_enter {
                 # on entering the thread, we set an VTALRM handler to cede
                 $SIG{VTALRM} = sub { cede };
                 # and then start the interval timer
                 Time::HiRes::setitimer &Time::HiRes::ITIMER_VIRTUAL, 0.01, 0.01;
              }; 
              Coro::on_leave {
                 # on leaving the thread, we stop the interval timer again
                 Time::HiRes::setitimer &Time::HiRes::ITIMER_VIRTUAL, 0, 0;
              }; 

              &{+shift};
           }  

           # use like this:
           timeslice {
              # The following is an endless loop that would normally
              # monopolise the process. Since it runs in a timesliced
              # environment, it will regularly cede to other threads.
              while () { }
           };

    killall
        Kills/terminates/cancels all coros except the currently running one.

        Note that while this will try to free some of the main interpreter

Changes
Revision history for Perl extension Coro.

TODO: should explore PerlIO::coroaio (perl leaks like hell).
TODO: maybe implement a default message channel, very much like Erlang's
      actor model (which is cool in a lot of important aspects (failures!),
      but very lacking in others (higher level ipc)).
TODO: unready_all
TODO: myhttpd header parsing
TODO: channel->maxsize(newsize)?

5.15  Tue Jun 30 10:28:06 CEST 2009
	- deprecate Coro::Socket, document how to get ipv6 support via
          AnyEvent::Socket instead.
	- implement signal->wait ($cb) interface, similar to semaphores.
        - work around SvOK not supporting getmagic, so we have to getmagic
          to test for undef :/ (reported by Matthias Waldorf).
        - load Coro::AnyEvent in all modules using AnyEvent.
        - work around perl corrupting our internal data structures,
          reported by Tokuhiro Matsuno.
        - enable per-coroutine real and cpu time gathering
          (Coro::State::enable_times).

5.14  Wed Jun 24 01:37:48 CEST 2009
	- provide explicit functions to "cede" to the event loop in Coro::AnyEvent,
          as this seems to have been a difficult concept (poll, sleep, idle,
          idle_upto).
        - add Coro::AnyEvent::readable/writable functions.
	- clarify Coro::EV/Event/AnyEvent manpages.
        - free per-thread global scalars in the thread calling ->cancel, to
          avoid crashes when $_, $@ etc., are magical but some of those
          had already been freed.
        - "unexperimentalise" the callback interface for Coro::Semaphore.
	- speed up ready queue management/context switching by using a linked
          list instead of an array (~5-10%).
        - implement "watch" command in Coro::Debug shells.
        - for fun, implement time-slicing as an example in the manpage.
        - if AnyEvent detects EV or Event, but we don't have Coro::EV or
          Coro::Event, use the normal AnyEvent handling instead of dieing
          (the same is true for Coro::Handle).
        - properly document Coro::EV::timed_io_once.
        - avoid unneccessary ->cancel calls in Coro::Handle.
        - maybe make it work on mingw32 with win32 backend
          (based on patch by Yasuhiro Matsumoto).

5.132 Fri May 29 09:00:39 CEST 2009
	- do not keep a reference to the argument itself in
          Coro::Semaphore::guard, as it could change later.
        - support SO_RCVBUF/SO_SNDBUF nonstandard Coro::Socket options,
          should support a prepare callback.

5.131 Mon Mar 16 23:20:37 CET 2009
	- implement and document Coro->suspend, Coro->resume.
        - fix Coro::Select implementation to not (often) close
          the passed file descriptors (testcase provided by pippijn).

5.13 Mon Dec 15 21:51:42 CET 2008
	- EXPERIMENTAL: implement dynamic winds (on_enter/on_leave).
	- don't set diehook to C<undef> but instead to NULL, to avoid
          spurious warnings.
        - fix a lot of bugs in Coro::SemaphoreSet.
        - Coro::SemaphoreSet will less often create a semaphore needlessly.
        - add Coro::SemaphoreSet::count and wait methods.
        - take advantage of the new Guard module.
        - deprecate Coro::guard.
        - try to fix the dreaded 01_unblock tests once more. I hate it when
          testsuites need more fixing than the code they are supposed to test.
        - croak in more cases when a required callback isn't resolvable.
        - fix some minor issues in Coro::State->call/eval.
        - use current coroutine context instead of a temporary one
          when temporarily switching to another coroutine.
        - do not call C-level on_destroy handlers during global destruction,
          to avoid needless segfaults.

5.12 Sun Dec  7 13:30:38 CET 2008
        - add default config for MirOS, which seems to be a bug-to-bug
          compatible fork of openbsd ("world domination by releasing
          openbsd cvs before the openbsd folks do it" or so :).
        - free_padlist did destroy the names pad, not good, but didn't
          seem to bother perl - this could fix issues such as eval ""
          inside a function called from multiple coroutines.
        - use a different method to detect destruction time.
        - be more careful when freeing padlists just before global
          destruction.
        - fixed and expanded the call/cc example.
        - renamed _terminate to _coro_run.
        - new method Coro::Channel->shutdown.
        - try pthreads on openbsd <4.4 (broken sigaltstack, will
          pthreads fare better?).
        - be less picky about destroying "a" running coroutine.

5.11 Tue Nov 25 21:49:05 CET 2008
	- DEBUGGING in 5.10.0 is a mess: it constantly flags perfectly
          working code with failed assertions, introducing more bugs than
          it fixes, requiring elaborate workarounds :(

5.1  Mon Nov 24 08:54:59 CET 2008
	- wrote a small introductory tutorial - Coro::Intro.
	- convert Coro::Timer, Coro::Select and Coro::Util to rouse API.
        - Coro::Select did errornously dup the file descriptors
          and didn't work with all AnyEvent backends.
        - Coro::Select wasn't imported correctly form Coro::LWP, causing blocking
          LWP data transfers.
        - disassociate c contexts from coro objects - this is agruably more
          correct, but mostly allows sharing of cctxs between coro and state
          objects, for added memory savings and speed increases.
        - bumped $Coro::POOL_RSS up to 32kb by default.
        - no longer set the optype to OP_CUSTOM, as B::* understandably
          doesn't like this very much (and we *are* a type of entersub).
        - implement state cloning, just to prove that call/cc can be done.
        - automatically load Coro::AnyEvent in Coro::Handle.
        - wrap ->cancel calls in eval inside Coro::Handle as EV watchers
          do not have this method (and don't need it either).
        - speed up generic anyevent methods in Coro::Handle by using rouse

Coro/Debug.pm
=head1 NAME

Coro::Debug - various functions that help debugging Coro programs

=head1 SYNOPSIS

 use Coro::Debug;

 our $server = new_unix_server Coro::Debug "/tmp/socketpath";

 $ socat readline unix:/tmp/socketpath

=head1 DESCRIPTION

This module is an L<AnyEvent> user, you need to make sure that you use and
run a supported event loop.

This module provides some debugging facilities. Most will, if not handled
carefully, severely compromise the security of your program, so use it
only for debugging (or take other precautions).

It mainly implements a very primitive debugger that is evry easy to
integrate in your program:

   our $server = new_unix_server Coro::Debug "/tmp/somepath";
   # see new_unix_server, below, for more info

It lets you list running coroutines:

            state (rUnning, Ready, New or neither)
            |cctx allocated
            ||  resident set size (octets)
            ||  |   scheduled this many times
   > ps     ||  |   |
        PID SC  RSS USES Description              Where
   14572344 UC  62k 128k [main::]                 [dm-support.ext:47]
   14620056 -- 2260   13 [coro manager]           [Coro.pm:358]
   14620128 -- 2260  166 [unblock_sub scheduler]  [Coro.pm:358]
   17764008 N-  152    0 [EV idle process]        -
   13990784 -- 2596  10k timeslot manager         [cf.pm:454]
   81424176 --  18k 4758 [async pool idle]        [Coro.pm:257]
   23513336 -- 2624    1 follow handler           [follow.ext:52]
   40548312 --  15k 5597 player scheduler         [player-scheduler.ext:13]
   29138032 -- 2548  431 music scheduler          [player-env.ext:77]
   43449808 -- 2260 3493 worldmap updater         [item-worldmap.ext:115]
   33352488 --  19k 2845 [async pool idle]        [Coro.pm:257]
   81530072 --  13k  43k map scheduler            [map-scheduler.ext:65]
   30751144 --  15k 2204 [async pool idle]        [Coro.pm:257]

Lets you do backtraces on about any coroutine:

   > bt 18334288
   coroutine is at /opt/cf/ext/player-env.ext line 77
           eval {...} called at /opt/cf/ext/player-env.ext line 77
           ext::player_env::__ANON__ called at -e line 0
           Coro::_run_coro called at -e line 0

Or lets you eval perl code:

   > 5+7
   12

Or lets you eval perl code within other coroutines:

   > eval 18334288 caller(1); $DB::args[0]->method
   1

It can also trace subroutine entry/exits for most coroutines (those not
having recursed into a C function), resulting in output similar to:

   > loglevel 5
   > trace 94652688
   2007-09-27Z20:30:25.1368 (5) [94652688] enter Socket::sockaddr_in with (8481,\x{7f}\x{00}\x{00}\x{01})
   2007-09-27Z20:30:25.1369 (5) [94652688] leave Socket::sockaddr_in returning (\x{02}\x{00}...)
   2007-09-27Z20:30:25.1370 (5) [94652688] enter Net::FCP::Util::touc with (client_get)
   2007-09-27Z20:30:25.1371 (5) [94652688] leave Net::FCP::Util::touc returning (ClientGet)
   2007-09-27Z20:30:25.1372 (5) [94652688] enter AnyEvent::Impl::Event::io with (AnyEvent,fh,GLOB(0x9256250),poll,w,cb,CODE(0x8c963a0))
   2007-09-27Z20:30:25.1373 (5) [94652688] enter Event::Watcher::__ANON__ with (Event,poll,w,fd,GLOB(0x9256250),cb,CODE(0x8c963a0))
   2007-09-27Z20:30:25.1374 (5) [94652688] enter Event::io::new with (Event::io,poll,w,fd,GLOB(0x9256250),cb,CODE(0x8c963a0))
   2007-09-27Z20:30:25.1375 (5) [94652688] enter Event::Watcher::init with (Event::io=HASH(0x8bfb120),HASH(0x9b7940))

If your program uses the Coro::Debug::log facility:

   Coro::Debug::log 0, "important message";
   Coro::Debug::log 9, "unimportant message";

Then you can even receive log messages in any debugging session:

   > loglevel 5
   2007-09-26Z02:22:46 (9) unimportant message

Other commands are available in the shell, use the C<help> command for a list.

=head1 FUNCTIONS

None of the functions are being exported.

=over 4

=cut

package Coro::Debug;

use strict qw(subs vars);
no warnings;

use overload ();

use Carp ();
use Time::HiRes ();
use Scalar::Util ();

use Guard;

use AnyEvent ();
use AnyEvent::Util ();
use AnyEvent::Socket ();

use Coro ();
use Coro::Handle ();
use Coro::State ();
use Coro::AnyEvent ();
use Coro::Timer ();

our $VERSION = 5.15;

our %log;
our $SESLOGLEVEL = exists $ENV{PERL_CORO_DEFAULT_LOGLEVEL} ? $ENV{PERL_CORO_DEFAULT_LOGLEVEL} : -1;
our $ERRLOGLEVEL = exists $ENV{PERL_CORO_STDERR_LOGLEVEL}  ? $ENV{PERL_CORO_STDERR_LOGLEVEL}  : -1;

sub find_coro {
   my ($pid) = @_;

   if (my ($coro) = grep $_ == $pid, Coro::State::list) {
      $coro
   } else {
      print "$pid: no such coroutine\n";
      undef
   }
}

sub format_msg($$) {
   my ($time, $micro) = Time::HiRes::gettimeofday;
   my ($sec, $min, $hour, $day, $mon, $year) = gmtime $time;
   my $date = sprintf "%04d-%02d-%02dZ%02d:%02d:%02d.%04d",
                      $year + 1900, $mon + 1, $day, $hour, $min, $sec, $micro / 100;
   sprintf "%s (%d) %s", $date, $_[0], $_[1]
}

sub format_num4($) {
   my ($v) = @_;

   return sprintf "%4d"   , $v                     if $v <  1e4;
   # 1e5 redundant
   return sprintf "%3.0fk", $v /             1_000 if $v <  1e6;
   return sprintf "%1.1fM", $v /         1_000_000 if $v <  1e7 * .995;
   return sprintf "%3.0fM", $v /         1_000_000 if $v <  1e9;
   return sprintf "%1.1fG", $v /     1_000_000_000 if $v < 1e10 * .995;
   return sprintf "%3.0fG", $v /     1_000_000_000 if $v < 1e12;
   return sprintf "%1.1fT", $v / 1_000_000_000_000 if $v < 1e13 * .995;
   return sprintf "%3.0fT", $v / 1_000_000_000_000 if $v < 1e15;

   "++++"
}

=item log $level, $msg

Log a debug message of the given severity level (0 is highest, higher is
less important) to all interested parties.

=item stderr_loglevel $level

Set the loglevel for logging to stderr (defaults to the value of the
environment variable PERL_CORO_STDERR_LOGLEVEL, or -1 if missing).

=item session_loglevel $level

Set the default loglevel for new coro debug sessions (defaults to the
value of the environment variable PERL_CORO_DEFAULT_LOGLEVEL, or -1 if
missing).

=cut

sub log($$) {
   my ($level, $msg) = @_;
   $msg =~ s/\s*$/\n/;
   $_->($level, $msg) for values %log;
   printf STDERR format_msg $level, $msg if $level <= $ERRLOGLEVEL;
}

sub session_loglevel($) {
   $SESLOGLEVEL = shift;
}

sub stderr_loglevel($) {
   $ERRLOGLEVEL = shift;
}

=item trace $coro, $loglevel

Enables tracing the given coroutine at the given loglevel. If loglevel is
omitted, use 5. If coro is omitted, trace the current coroutine. Tracing
incurs a very high runtime overhead.

It is not uncommon to enable tracing on oneself by simply calling
C<Coro::Debug::trace>.

A message will be logged at the given loglevel if it is not possible to
enable tracing.

=item untrace $coro

Disables tracing on the given coroutine.

=cut

sub trace {
   my ($coro, $loglevel) = @_;

   $coro ||= $Coro::current;
   $loglevel = 5 unless defined $loglevel;

   (Coro::async {
      if (eval { Coro::State::trace $coro, Coro::State::CC_TRACE | Coro::State::CC_TRACE_SUB; 1 }) {
         Coro::Debug::log $loglevel, sprintf "[%d] tracing enabled", $coro + 0;
         $coro->{_trace_line_cb} = sub {
            Coro::Debug::log $loglevel, sprintf "[%d] at %s:%d\n", $Coro::current+0, @_;
         };
         $coro->{_trace_sub_cb} = sub {
            Coro::Debug::log $loglevel, sprintf "[%d] %s %s %s\n",
               $Coro::current+0,
               $_[0] ? "enter" : "leave",
               $_[1],
               $_[2] ? ($_[0] ? "with (" : "returning (") . (
                  join ",",
                     map {
                        my $x = ref $_ ? overload::StrVal $_ : $_;
                        (substr $x, 40) = "..." if 40 + 3 < length $x;
                        $x =~ s/([^\x20-\x5b\x5d-\x7e])/sprintf "\\x{%02x}", ord $1/ge;
                        $x
                     } @{$_[2]}
               ) . ")" : "";
         };

         undef $coro; # the subs keep a reference which we do not want them to do
      } else {
         Coro::Debug::log $loglevel, sprintf "[%d] unable to enable tracing: %s", $Coro::current + 0, $@;
      }
   })->prio (Coro::PRIO_MAX);

   Coro::cede;
}

sub untrace {
   my ($coro) = @_;

   $coro ||= $Coro::current;

   (Coro::async {
      Coro::State::trace $coro, 0;
      delete $coro->{_trace_sub_cb};
      delete $coro->{_trace_line_cb};
   })->prio (Coro::PRIO_MAX);

   Coro::cede;
}

=item command $string

Execute a debugger command, sending any output to STDOUT. Used by
C<session>, below.

=cut

sub command($) {
   my ($cmd) = @_;

   $cmd =~ s/\s+$//;

   if ($cmd =~ /^ps (?:\s* (\S+))? $/x) {
      my $times = Coro::State::enable_times;
      my $flags = $1;
      my $verbose = $flags =~ /v/;
      my $desc_format = $flags =~ /w/ ? "%-24s" : "%-24.24s";
      my $tim0_format = $times ? " %9s %8s " : " ";
      my $tim1_format = $times ? " %9.3f %8.3f " : " ";
      my $buf = sprintf "%20s %s%s %4s %4s$tim0_format$desc_format %s\n",
                        "PID", "S", "C", "RSS", "USES",
                        $times ? ("t_real", "t_cpu") : (),
                        "Description", "Where";
      for my $coro (reverse Coro::State::list) {
         my @bt;
         Coro::State::call ($coro, sub {
            # we try to find *the* definite frame that gives msot useful info
            # by skipping Coro frames and pseudo-frames.
            for my $frame (1..10) {
               my @frame = caller $frame;
               @bt = @frame if $frame[2];
               last unless $bt[0] =~ /^Coro/;
            }
         });
         $bt[1] =~ s/^.*[\/\\]// if @bt && !$verbose;
         $buf .= sprintf "%20s %s%s %4s %4s$tim1_format$desc_format %s\n",
                         $coro+0,
                         $coro->is_new ? "N" : $coro->is_running ? "U" : $coro->is_ready ? "R" : "-",
                         $coro->is_traced ? "T" : $coro->has_cctx ? "C" : "-",
                         format_num4 $coro->rss,
                         format_num4 $coro->usecount,
                         $times ? $coro->times : (),
                         $coro->debug_desc,
                         (@bt ? sprintf "[%s:%d]", $bt[1], $bt[2] : "-");
      }
      print $buf;

   } elsif ($cmd =~ /^bt\s+(\d+)$/) {
      if (my $coro = find_coro $1) {
         my $bt;
         Coro::State::call ($coro, sub {
            local $Carp::CarpLevel = 2;
            $bt = eval { Carp::longmess "coroutine is" } || "$@";
         });
         if ($bt) {
            print $bt;
         } else {
            print "$1: unable to get backtrace\n";
         }
      }

   } elsif ($cmd =~ /^(?:e|eval)\s+(\d+)\s+(.*)$/) {
      if (my $coro = find_coro $1) {
         my $cmd = eval "sub { $2 }";
         my @res;
         Coro::State::call ($coro, sub { @res = eval { &$cmd } });
         print $@ ? $@ : (join " ", @res, "\n");
      }

   } elsif ($cmd =~ /^(?:tr|trace)\s+(\d+)$/) {
      if (my $coro = find_coro $1) {
         trace $coro;
      }

   } elsif ($cmd =~ /^(?:ut|untrace)\s+(\d+)$/) {
      if (my $coro = find_coro $1) {
         untrace $coro;
      }

   } elsif ($cmd =~ /^cancel\s+(\d+)$/) {
      if (my $coro = find_coro $1) {
         $coro->cancel;
      }

   } elsif ($cmd =~ /^ready\s+(\d+)$/) {
      if (my $coro = find_coro $1) {
         $coro->ready;
      }

   } elsif ($cmd =~ /^kill\s+(\d+)(?:\s+(.*))?$/) {
      my $reason = defined $2 ? $2 : "killed";

      if (my $coro = find_coro $1) {
         $coro->throw ($reason);
      }

   } elsif ($cmd =~ /^enable_times(\s+\S.*)?\s*$/) {
      my $enable = defined $1 ? 1*eval $1 : !Coro::State::enable_times;

      Coro::State::enable_times $enable;

      print "per-thread real and process time gathering ", $enable ? "enabled" : "disabled", ".\n";

   } elsif ($cmd =~ /^help$/) {
      print <<EOF;
ps [w|v]                show the list of all coroutines (wide, verbose)
bt <pid>                show a full backtrace of coroutine <pid>
eval <pid> <perl>       evaluate <perl> expression in context of <pid>
trace <pid>             enable tracing for this coroutine
untrace <pid>           disable tracing for this coroutine
kill <pid> <reason>	throws the given <reason> string in <pid>
cancel <pid>		cancels this coroutine
ready <pid>		force <pid> into the ready queue
enable_times <enable>	enable or disable time profiling in ps
<anything else>         evaluate as perl and print results
<anything else> &       same as above, but evaluate asynchronously
                        you can use (find_coro <pid>) in perl expressions
                        to find the coro with the given pid, e.g.
                        (find_coro 9768720)->ready
EOF

   } elsif ($cmd =~ /^(.*)&$/) {
      my $cmd = $1;
      my $sub = eval "sub { $cmd }";
      my $fh = select;
      Coro::async_pool {
         $Coro::current->{desc} = $cmd;
         my $t = Time::HiRes::time;
         my @res = eval { &$sub };
         $t = Time::HiRes::time - $t;
         print {$fh}
            "\rcommand: $cmd\n",
            "execution time: $t\n",
            "result: ", $@ ? $@ : (join " ", @res) . "\n",
            "> ";
      };

   } else {
      my @res = eval $cmd;
      print $@ ? $@ : (join " ", @res) . "\n";
   }
}

=item session $fh

Run an interactive debugger session on the given filehandle. Each line entered
is simply passed to C<command>.

=cut

sub session($) {
   my ($fh) = @_;

   $fh = Coro::Handle::unblock $fh;
   my $old_fh = select $fh;
   my $guard = guard { select $old_fh };

   my $loglevel = $SESLOGLEVEL;
   local $log{$Coro::current} = sub {
      return unless $_[0] <= $loglevel;
      print $fh "\015", (format_msg $_[0], $_[1]), "> ";
   };

   print "coro debug session. use help for more info\n\n";

   while ((print "> "), defined (my $cmd = $fh->readline ("\012"))) {
      if ($cmd =~ /^exit\s*$/) {
         print "bye.\n";
         last;

      } elsif ($cmd =~ /^(?:ll|loglevel)\s*(\d+)?\s*/) {
         $loglevel = defined $1 ? $1 : -1;

      } elsif ($cmd =~ /^(?:w|watch)\s*([0-9.]*)\s+(.*)/) {
         my ($time, $cmd) = ($1*1 || 1, $2);
         my $cancel;

         Coro::async {
            $Coro::current->{desc} = "watch $cmd";
            select $fh;
            until ($cancel) {
               command $cmd;
               Coro::Timer::sleep $time;
            }
         };

         $fh->readable;
         $cancel = 1;

      } elsif ($cmd =~ /^help\s*/) {
         command $cmd;
         print <<EOF;
loglevel <int>		enable logging for messages of level <int> and lower
watch <time> <command>  repeat the given command until STDIN becomes readable
exit			end this session
EOF
      } else {
         command $cmd;
      }

      Coro::cede;
   }
}

=item $server = new_unix_server Coro::Debug $path

Creates a new unix domain socket that listens for connection requests and
runs C<session> on any connection. Normal unix permission checks and umask
applies, so you can protect your socket by puttint it into a protected
directory.

The C<socat> utility is an excellent way to connect to this socket:

   socat readline /path/to/socket

Socat also offers history support:

   socat readline:history=/tmp/hist.corodebug /path/to/socket

The server accepts connections until it is destroyed, so you must keep
the return value around as long as you want the server to stay available.

=cut

sub new_unix_server {
   my ($class, $path) = @_;

   unlink $path;
   my $unlink_guard = guard { unlink $path };

   AnyEvent::Socket::tcp_server "unix/", $path, sub {
      my ($fh) = @_;
      $unlink_guard; # mention it
      Coro::async_pool {
         $Coro::current->desc ("[Coro::Debug session]");
         session $fh;
      };
   } or Carp::croak "Coro::Debug::new_unix_server($path): $!";
}

=item $server = new_tcp_server Coro::Debug $port

Similar to C<new_unix_server>, but binds on a TCP port. I<Note that this is
usually results in a gaping security hole>.

Currently, only a TCPv4 socket is created, in the future, a TCPv6 socket
might also be created.

=cut

sub new_tcp_server {
   my ($class, $port) = @_;

   AnyEvent::Socket::tcp_server undef, $port, sub {
      my ($fh) = @_;
      Coro::async_pool {
         $Coro::current->desc ("[Coro::Debug session]");
         session $fh;
      };
   } or Carp::croak "Coro::Debug::new_tcp_server($port): $!";
}

sub DESTROY {
   my ($self) = @_;

   unlink $self->{path} if exists $self->{path};
   %$self = ();
}

1;

=back

=head1 AUTHOR

 Marc Lehmann <schmorp@schmorp.de>
 http://home.schmorp.de/

=cut


